dstackai/dstack
# Support dstack volumes

Closes #1158 

This PR adds support for network volumes in dstack.

* Introduces `volume` configuration.
* Implements `dstack apply`, `dstack delete`, `dstack volume list` for volumes.
* Adds async API for managing volumes and background processing task to process volumes.
* Extends shim capabilities to format and mount volumes on the host and mount volumes in the container.
* Extends Compute interface to support volumes.
* Implements dstack-managed and external volumes for AWS.
* Updates docs reference with volumes.
* MInor refactoring.

Notes:
* Both external and dstack-managed volumes are supported.
* Attaching multiple volumes per run is supported.
* Volumes can be used with instances in pools – they are attached/detached automatically with job start/end.
* There is no way to see volume attachments/mount via CLI.
* No way to manually attach/detach volumes.

Test volumes:

1. Create a volume with `dstack apply`:

```
type: volume
name: my-dstack-aws-volume
backend: aws
region: eu-central-1
size: 100GB
```

2. Run a dev environment with the volume attached:

```
type: dev-environment
ide: vscode
volumes:
  - name: my-dstack-aws-volume
    path: /my_volume_data
```

3. Save data on the volume, stop the run, run the dev env again and see the volume data persist.
